windows/code/RNDeviceInfoCPP.h

```diff
@@ -75,6 +121,53 @@ namespace winrt::RNDeviceInfoCPP
         });
     }
 
+    REACT_SYNC_METHOD(isKeyboardConnectedSync);
+    bool isKeyboardConnectedSync() noexcept
+    {
+      auto keyboardCapabilities = winrt::Windows::Devices::Input::KeyboardCapabilities();
+      return keyboardCapabilities.KeyboardPresent();
+    }
+
+    REACT_METHOD(isKeyboardConnected);
+    void isKeyboardConnected(ReactPromise<bool> promise) noexcept
+    {
+      promise.Resolve(isKeyboardConnectedSync());
+    }
+
+    REACT_SYNC_METHOD(isMouseConnectedSync);
+    bool isMouseConnectedSync() noexcept
+    {
+      auto mouseCapabilities = winrt::Windows::Devices::Input::MouseCapabilities();
+      return mouseCapabilities.MousePresent();
+    }
+
+    REACT_METHOD(isMouseConnected);
+    void isMouseConnected(ReactPromise<bool> promise) noexcept
+    {
+      promise.Resolve(isMouseConnectedSync());
+    }
+
+    REACT_METHOD(isTabletMode);
+    void isTabletMode(ReactPromise<bool> promise) noexcept
+    {
+      m_reactContext.UIDispatcher().Post([promise]() {
+        auto view = winrt::Windows::UI::ViewManagement::UIViewSettings::GetForCurrentView();
```

Comment:
you'll want to caveat this with "if not running inside an islands app" since this will crash outside UWP. In those cases, you'll want to use the APIs I mentioned offline.

Reply:
What is an Islands app, and I will look into adding those others for isTabletMode as well.  
When is RNW not running inside UWP?

Reply:
XAML Islands (aka Desktop bridge) is a set of technologies that allow win32 apps (i.e. non-appcontainer) to have XAML UI in them; see [this link](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands).
For RNW, it is important that we support this scenario for certain consumers (as in, things might dim down but should not crash), ideally we respond to the right win32-style events via window messages instead, for example. There are helpers to subscribe to window messages from a native module if you `#include <DesktopWindowBridge.h>`  in RNW, there's an example of it in the inbox [DeviceInfo module](https://github.com/microsoft/react-native-windows/blob/fe6d6a12c143267eb2a84a89cc417b1c43221006/vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp#L69-L84)

We have a Win32/RNW island app in the repo in `\packages\playground\windows\playground-win32.sln`

Reply:
So all of this is completely new and foreign to me, first time hearing of any of it.

So what I need to do here is add a 
`if (IsWinUI3Island())` and this particular UI:ViewManagement I can't use in an island?
And then use the other APIs otherwise?

Reply:
yes that's right. There are helpers to detect when you're in an island, see [helpers.h](https://github.com/microsoft/react-native-windows/blob/fe6d6a12c143267eb2a84a89cc417b1c43221006/vnext/Microsoft.ReactNative/Utils/Helpers.h#L33)